src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -106,13 +107,17 @@ void ShenandoahAdaptiveHeuristics::choose_collection_set_from_regiondata(Shenand
   // particular, regions that have reached tenure age will be sorted into this array before younger regions that contain
   // more garbage.  This represents one of the reasons why we keep looking at regions even after we decide, for example,
   // to exclude one of the regions because it might require evacuation of too much live data.
-
+  bool is_generational = heap->mode()->is_generational();
   for (size_t idx = 0; idx < size; idx++) {
     ShenandoahHeapRegion* r = data[idx]._region;
-    size_t biased_garbage = data[idx]._garbage;
-
-    size_t new_cset    = cur_cset + r->get_live_data_bytes();
-
+    size_t new_cset;
+    if (is_generational && (r->age() >= InitialTenuringThreshold)) {
```

Comment:
Are we changing region affiliation en masse?

Reply:
We don't "promote" regions by relabeling region affiliation.  But I found that the computed budget for promotion reserve is way too low in the case that entire regions have reached tenure age.  So I made an adjustment to compute larger promotion reserves when we know that all live data within a collection set region is known to be promoted.